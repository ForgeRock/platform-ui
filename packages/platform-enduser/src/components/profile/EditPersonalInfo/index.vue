<!-- Copyright 2019-2020 ForgeRock AS. All Rights Reserved

Use of this code requires a commercial software license with ForgeRock AS.
or with one of its affiliates. All use shall be exclusively subject
to such license between the licensee and ForgeRock AS. -->
<template>
  <BModal
    id="userDetailsModal"
    modal-class="fr-full-screen"
    ref="fsModal"
    cancel-variant="outline-secondary"
    @show="setModal"
    @keydown.enter.native.prevent="saveForm">
    <template v-slot:modal-header>
      <div class="d-flex w-100 h-100">
        <h5 class="modal-title align-self-center text-center">
          {{ title }}
        </h5>
        <button
          type="button"
          aria-label="Close"
          class="close"
          @click="hideModal">
          <i class="material-icons-outlined font-weight-bolder md-24">
            close
          </i>
        </button>
      </div>
    </template>

    <!-- Editing profile currently only supports String, Number and Boolean-->
    <BContainer>
      <BRow>
        <BCol
          sm="8"
          offset-sm="2">
          <BForm
            style="flex-direction: column;"
            v-if="formFields.length > 0"
            class="mb-3 fr-edit-personal-form"
            name="edit-personal-form">
            <ValidationObserver
              ref="observer"
              v-slot:default>
              <template v-for="(field, index) in formFields">
                <BFormGroup
                  style="min-width: 200px;"
                  :key="index"
                  v-if="field.type === 'string' || field.type === 'number' || field.type === 'boolean'">
                  <FrField
                    :field="field"
                    :prepend-title="true" />
                </BFormGroup>
              </template>
            </ValidationObserver>
          </BForm>
          <template v-else>
            <h3 class="text-center">
              {{ $t('pages.profile.editProfile.noFields') }}
            </h3>
          </template>
        </BCol>
      </BRow>
    </BContainer>
    <template v-slot:modal-footer="{ cancel }">
      <BButton
        variant="outline-secondary mr-2"
        @click="cancel()">
        {{ $t('common.cancel') }}
      </BButton>
      <BButton
        variant="primary"
        :disabled="internalUser"
        @click="saveForm">
        {{ $t('common.saveChanges') }}
      </BButton>
    </template>
  </BModal>
</template>

<script>
import {
  cloneDeep,
  each,
  filter,
  map,
} from 'lodash';
import { mapState } from 'vuex';
import ResourceMixin from '@forgerock/platform-shared/src/mixins/ResourceMixin';
import RestMixin from '@forgerock/platform-shared/src/mixins/RestMixin';
import NotificationMixin from '@forgerock/platform-shared/src/mixins/NotificationMixin';
import FrField from '@forgerock/platform-shared/src/components/Field';

/**
 * @description Displays a users profile, auto generates fields based off of resource schema. Currently only displays strings, numbers and booleans. In the case of a policy
 * save error it will highlight the appropriate field and display a policy error. For custom profile changes (e.g. adding a dropdown) this would be the primary file to add these
 * adjustments.
 */
export default {
  name: 'EditPersonalInfo',
  mixins: [
    ResourceMixin,
    RestMixin,
    NotificationMixin,
  ],
  components: {
    FrField,
  },
  computed: {
    ...mapState({
      userId: (state) => state.UserStore.userId,
      managedResource: (state) => state.UserStore.managedResource,
      internalUser: (state) => state.UserStore.internalUser,
    }),
  },
  props: {
    schema: { type: Object, required: true },
    profile: { type: Object, required: true },
    autoOpen: { type: Boolean, required: false, default: false },
  },
  data() {
    return {
      formFields: [],
      originalFormFields: [],
      title: this.$t('pages.profile.editProfile.userDetailsTitle'),
    };
  },
  mounted() {
    if (this.autoOpen) {
      this.$root.$emit('bv::show::modal', 'userDetailsModal');
    }
  },
  methods: {
    generateFormFields() {
      const { order, properties, required } = this.schema;
      const filteredOrder = filter(order, (propName) => properties[propName].viewable
                            && properties[propName].userEditable
                            && properties[propName].type !== 'array'
                            && properties[propName].type !== 'object');
      const formFields = map(filteredOrder, (name) => ({
        name,
        key: name,
        title: `${properties[name].title} ${required.includes(name) ? '' : this.$t('pages.profile.editProfile.optional')}`,
        value: this.profile[name] || null,
        type: properties[name].type,
        validation: required.includes(name) ? 'required' : '',
      }));

      return formFields;
    },
    hideModal() {
      this.$refs.fsModal.hide();
    },
    setModal() {
      const formFields = this.generateFormFields();

      this.formFields = formFields;
      this.originalFormFields = cloneDeep(formFields);
    },
    async saveForm() {
      const isValid = await this.$refs.observer.validate();
      if (isValid) {
        const idmInstance = this.getRequestService();
        const policyFields = {};

        each(this.formFields, (field) => {
          if (field.value !== null) {
            policyFields[field.name] = field.value;
          }
        });

        idmInstance.post(`policy/${this.managedResource}/${this.userId}?_action=validateObject`, policyFields).then((policyResult) => {
          if (policyResult.data.failedPolicyRequirements.length === 0) {
            this.$emit('updateProfile', this.generateUpdatePatch(this.originalFormFields, this.formFields));
            this.$refs.observer.reset();
            this.hideModal();
          } else {
            const generatedErrors = this.findPolicyError({
              data: {
                detail: {
                  failedPolicyRequirements: policyResult.data.failedPolicyRequirements,
                },
              },
            }, this.formFields);

            this.$refs.observer.reset();
            if (generatedErrors.length > 0) {
              each(generatedErrors, (generatedError) => {
                if (generatedError.exists) {
                  this.$refs.observer.setErrors({
                    [generatedError.field]: [generatedError.msg],
                  });
                }
              });
            } else {
              this.displayNotification('IDMMessages', 'error', this.$t('pages.profile.editProfile.failedProfileSave'));
            }
          }
        });
      }
    },
  },
};
</script>
