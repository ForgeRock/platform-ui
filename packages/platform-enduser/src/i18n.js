/**
 * Copyright (c) 2020-2024 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import { createI18n } from 'vue-i18n';
import { merge } from 'lodash';
import dateTimeFormats from '@forgerock/platform-shared/src/dateTimeFormats.json';

function loadLocaleMessages() {
  const locales = require.context('./locales', true, /[A-Za-z0-9-_,\s]+\.json$/i);
  // Can't use workspace symlink due to jest issues
  const sharedLocales = require.context('../../platform-shared/src/locales', true, /[A-Za-z0-9-_,\s]+\.json$/i);
  const messages = {};

  locales.keys().forEach((key) => {
    const matched = key.match(/([A-Za-z0-9-_]+)\./i);
    if (matched && matched.length > 1) {
      const locale = matched[1];
      let generatedTranslation = locales(key);

      // Check for shared local and merge in if it exists
      const sharedLocaleExists = sharedLocales.keys().includes(key);
      if (sharedLocaleExists) {
        generatedTranslation = merge(sharedLocales(key), generatedTranslation);
      }

      messages[locale] = generatedTranslation;
    }
  });

  return messages;
}

export default createI18n({
  messages: loadLocaleMessages(),
  silentFallbackWarn: true,
  silentTranslationWarn: true,
  dateTimeFormats,
});
