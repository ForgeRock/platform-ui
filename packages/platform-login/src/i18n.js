/**
 * Copyright (c) 2020-2021 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import Vue from 'vue';
import VueI18n from 'vue-i18n';
import { merge } from 'lodash';
import dateTimeFormats from '@forgerock/platform-shared/src/dateTimeFormats.json';

Vue.use(VueI18n);

function loadLocaleMessages() {
  const locales = require.context('./locales', true, /[A-Za-z0-9-_,\s]+\.json$/i);
  // Can't use workspace symlink due to jest issues
  const sharedLocales = require.context('../../platform-shared/src/locales', true, /[A-Za-z0-9-_,\s]+\.json$/i);
  const messages = {};

  locales.keys().forEach((key) => {
    const matched = key.match(/([A-Za-z0-9-_]+)\./i);
    if (matched && matched.length > 1) {
      const locale = matched[1];
      let generatedTranslation = locales(key);

      // Check for shared local and merge in if it exists
      const sharedLocaleExists = sharedLocales.keys().includes(key);
      if (sharedLocaleExists) {
        generatedTranslation = merge(sharedLocales(key), generatedTranslation);
      }

      messages[locale] = generatedTranslation;
    }
  });

  return messages;
}

Vue.mixin({
  methods: {
    translationExists(path) {
      return this.$t(path) !== path;
    },
  },
});

export default new VueI18n({
  locale: process.env.VUE_APP_I18N_LOCALE || 'en',
  fallbackLocale: process.env.VUE_APP_I18N_FALLBACK_LOCALE || 'en',
  messages: loadLocaleMessages(),
  silentTranslationWarn: true,
  dateTimeFormats,
});
