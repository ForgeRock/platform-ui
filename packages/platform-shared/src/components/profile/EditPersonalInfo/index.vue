<!-- Copyright (c) 2020-2021 ForgeRock. All rights reserved.

This software may be modified and distributed under the terms
of the MIT license. See the LICENSE file for details. -->
<template>
  <ValidationObserver
    v-slot="{ invalid }"
    ref="observer">
    <BModal
      id="userDetailsModal"
      ref="fsModal"
      size="lg"
      cancel-variant="outline-secondary"
      @show="setModal"
      @keydown.enter.native.prevent="saveForm">
      <template #modal-header>
        <div class="d-flex w-100 h-100">
          <h5 class="modal-title align-self-center text-center">
            {{ title }}
          </h5>
          <button
            type="button"
            :aria-label="$t('common.close')"
            class="close"
            @click="hideModal">
            <FrIcon
              class="font-weight-bolder md-24"
              name="close" />
          </button>
        </div>
      </template>
      <!-- Editing profile currently only supports String, Number and Boolean-->
      <BContainer>
        <BRow>
          <template v-if="formFields.length">
            <template v-for="(field, index) in formFields">
              <BFormGroup
                class="w-100"
                :key="index"
                :label="title"
                label-sr-only
                v-if="field.type === 'string' || field.type === 'number' || field.type === 'boolean'">
                <FrField
                  v-model="field.value"
                  :label="field.title"
                  :name="field.name"
                  :type="field.type"
                  :validation="field.validation" />
              </BFormGroup>
              <FrListField
                v-else-if="field.type === 'array' && field.name !== 'privileges'"
                v-model="field.value"
                class="w-100"
                :key="index"
                :description="field.description"
                :items="field.items"
                :label="field.title"
                :name="field.name"
                :required="field.required"
                :index="index"
                v-on="$listeners"
                @input="updateField(index, $event)" />
            </template>
          </template>
          <h3
            v-else
            class="text-center">
            {{ $t('profile.editProfile.noFields') }}
          </h3>
        </BRow>
      </BContainer>
      <template #modal-footer="{ cancel }">
        <BButton
          variant="link mr-2"
          @click="cancel()">
          {{ $t('common.cancel') }}
        </BButton>
        <BButton
          variant="primary"
          :disabled="internalUser || invalid"
          @click="saveForm">
          {{ $t('common.save') }}
        </BButton>
      </template>
    </BModal>
  </ValidationObserver>
</template>

<script>
import {
  cloneDeep,
  each,
  filter,
  map,
  reject,
} from 'lodash';
import {
  BButton,
  BContainer,
  BFormGroup,
  BModal,
  BRow,
} from 'bootstrap-vue';
import { ValidationObserver } from 'vee-validate';
import { mapState } from 'vuex';
import ResourceMixin from '@forgerock/platform-shared/src/mixins/ResourceMixin';
import RestMixin from '@forgerock/platform-shared/src/mixins/RestMixin';
import NotificationMixin from '@forgerock/platform-shared/src/mixins/NotificationMixin';
import FrField from '@forgerock/platform-shared/src/components/Field';
import FrIcon from '@forgerock/platform-shared/src/components/Icon';
import FrListField from '@forgerock/platform-shared/src/components/ListField';
import ListsMixin from '@forgerock/platform-shared/src/mixins/ListsMixin';

/**
 * @description Displays a users profile, auto generates fields based off of resource schema. Currently only displays strings, numbers and booleans. In the case of a policy
 * save error it will highlight the appropriate field and display a policy error. For custom profile changes (e.g. adding a dropdown) this would be the primary file to add these
 * adjustments.
 */
export default {
  name: 'EditPersonalInfo',
  mixins: [
    ListsMixin,
    NotificationMixin,
    ResourceMixin,
    RestMixin,
  ],
  components: {
    BButton,
    BContainer,
    BFormGroup,
    BModal,
    BRow,
    FrField,
    FrIcon,
    FrListField,
    ValidationObserver,
  },
  computed: {
    ...mapState({
      userId: (state) => state.UserStore.userId,
      managedResource: (state) => state.UserStore.managedResource,
      internalUser: (state) => state.UserStore.internalUser,
    }),
  },
  props: {
    /**
     * Schema data for profile
     */
    schema: {
      type: Object,
      required: true,
    },
    /**
     * Profile data
     */
    profile: {
      type: Object,
      required: true,
    },
  },
  data() {
    return {
      formFields: [],
      originalFormFields: [],
      title: this.$t('pages.profile.editProfile.userDetailsTitle'),
    };
  },
  methods: {
    /**
     * Get form fields from schema and profile data
     *
     * @returns {Object} profile fields
     */
    generateFormFields() {
      const { order, properties, required } = this.schema;
      const filteredOrder = filter(order, (propName) => properties[propName].userEditable
                            && properties[propName].viewable
                            && properties[propName].scope !== 'private'
                            && properties[propName].type !== 'object');
      const formFields = map(filteredOrder, (name) => ({
        name,
        title: `${properties[name].title} ${required.includes(name) ? '' : this.$t('pages.profile.editProfile.optional')}`,
        value: this.profile[name] || null,
        type: properties[name].type,
        description: properties[name].description,
        items: properties[name].items,
        validation: required.includes(name) ? 'required' : '',
      }));

      return formFields;
    },
    /**
     * Hide the modal
     */
    hideModal() {
      this.$refs.fsModal.hide();
    },
    /**
     * Set the form fields for the modal. Maintain the original values for patch
     */
    setModal() {
      const formFields = this.generateFormFields();

      this.formFields = formFields;
      this.originalFormFields = cloneDeep(formFields);
    },
    /**
     * Save the updated profile information via patch.
     * Values are validated by backend before saving.
     */
    async saveForm() {
      const isValid = await this.$refs.observer.validate();
      if (isValid) {
        const idmInstance = this.getRequestService();
        const policyFields = {};

        each(this.formFields, (field) => {
          if (field.value !== null) {
            policyFields[field.name] = field.value;
          }
        });

        idmInstance.post(`policy/${this.managedResource}/${this.userId}?_action=validateObject`, policyFields).then((policyResult) => {
          // reject any failedPolicyRequirements on properties that don't exist in this.formFields
          policyResult.data.failedPolicyRequirements = reject(policyResult.data.failedPolicyRequirements, (policy) => !map(this.formFields, 'name').includes(policy.property));

          if (policyResult.data.failedPolicyRequirements.length === 0) {
            this.$emit('updateProfile', this.generateUpdatePatch(this.originalFormFields, this.formFields));
            this.$refs.observer.reset();
            this.hideModal();
          } else {
            const generatedErrors = this.findPolicyError({
              data: {
                detail: {
                  failedPolicyRequirements: policyResult.data.failedPolicyRequirements,
                },
              },
            }, this.formFields);

            this.$refs.observer.reset();
            if (generatedErrors.length > 0) {
              each(generatedErrors, (generatedError) => {
                if (generatedError.exists) {
                  this.$refs.observer.setErrors({
                    [generatedError.field]: [generatedError.msg],
                  });
                }
              });
            } else {
              this.displayNotification('IDMMessages', 'error', this.$t('pages.profile.failedProfileSave'));
            }
          }
        });
      }
    },
    /**
     * Update form data model with a new value. Used for editing a list field
     *
     * @param {Number} index index of form field
     * @param {Array} newValue new value of form field
     */
    updateField(index, newValue) {
      this.formFields[index].value = newValue;
      this.$forceUpdate();
    },
  },
};
</script>
