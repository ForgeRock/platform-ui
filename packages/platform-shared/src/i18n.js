/**
 * Copyright 2019 ForgeRock AS. All Rights Reserved
 *
 * Use of this code requires a commercial software license with ForgeRock AS.
 * or with one of its affiliates. All use shall be exclusively subject
 * to such license between the licensee and ForgeRock AS.
 */
import Vue from 'vue';
import VueI18n from 'vue-i18n';
import dateTimeFormats from './dateTimeFormats.json';

Vue.use(VueI18n);

function loadLocaleMessages() {
  const locales = require.context('./locales', true, /[A-Za-z0-9-_,\s]+\.json$/i);
  const messages = {};

  locales.keys().forEach((key) => {
    const matched = key.match(/([A-Za-z0-9-_]+)\./i);
    if (matched && matched.length > 1) {
      const locale = matched[1];
      messages[locale] = locales(key);
    }
  });

  return messages;
}

export default new VueI18n({
  locale: process.env.VUE_APP_I18N_LOCALE || 'en',
  fallbackLocale: process.env.VUE_APP_I18N_FALLBACK_LOCALE || 'en',
  messages: loadLocaleMessages(),
  silentTranslationWarn: true,
  dateTimeFormats,
});
