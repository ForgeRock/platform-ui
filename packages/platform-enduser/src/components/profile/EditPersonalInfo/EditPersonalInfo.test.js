/**
 * Copyright 2019 ForgeRock AS. All Rights Reserved
 *
 * Use of this code requires a commercial software license with ForgeRock AS.
 * or with one of its affiliates. All use shall be exclusively subject
 * to such license between the licensee and ForgeRock AS.
 */
import BootstrapVue from 'bootstrap-vue';
import { createLocalVue, shallowMount } from '@vue/test-utils';
import { ValidationProvider, ValidationObserver } from 'vee-validate';
import _ from 'lodash';
import Vuex from 'vuex';
import i18n from '@/i18n';
import EditPersonalInfo from '@/components/profile/EditPersonalInfo';

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(BootstrapVue);

describe('EditPersonalInfo.vue', () => {
  const store = new Vuex.Store({});

  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(EditPersonalInfo, {
      localVue,
      i18n,
      store,
      stubs: {
        ValidationProvider,
        ValidationObserver,
      },
      propsData: {
        schema: {
          order: ['test'],
          properties: {
            test: {
              viewable: true,
              type: 'string',
              title: 'test title',
              userEditable: true,
            },
          },
          required: [],
        },
        profile: { test: 'test' },
      },
      computed: {
        userId: '1234',
        managedResource: {},
        internalUser: true,
      },
    });

    jest.spyOn(wrapper.vm, 'saveForm');
  });

  afterEach(() => {
    wrapper = null;
  });

  it('EditPersonalInfo modal loaded', () => {
    expect(wrapper.name()).toBe('EditPersonalInfo');
    expect(wrapper.isVisible()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders a title', () => {
    expect(wrapper.vm.title).toBe('Edit your personal info');
  });

  it('creates patches array correctly', () => {
    const original = [{
      name: 'description',
      value: null,
    }, {
      name: 'telephoneNumber',
      value: '123-456-7890',
    }, {
      name: 'city',
      value: 'Portland',
    }, {
      name: 'postalCode',
      value: null,
    }];

    const newForm = [{
      name: 'description',
      value: 'new description',
    }, {
      name: 'telephoneNumber',
      value: '',
    }, {
      name: 'city',
      value: 'Vancouver',
    }, {
      name: 'postalCode',
      value: null,
    }];

    const patches = wrapper.vm.generateUpdatePatch(original, newForm);

    expect(patches.length).toBe(3);
    expect(patches[0].operation).toBe('add');
    expect(patches[0].field).toBe('/description');
    expect(patches[0].value).toBe('new description');
    expect(patches[1].operation).toBe('remove');
    expect(patches[1].field).toBe('/telephoneNumber');
    expect(patches[2].operation).toBe('add');
    expect(patches[2].field).toBe('/city');
    expect(patches[2].value).toBe('Vancouver');
  });

  describe('#generateFormFields', () => {
    it('should create the proper fields based on schema', () => {
      const formFields = wrapper.vm.generateFormFields();
      const firstFormField = _.first(formFields);

      expect(typeof formFields).toBe('object');
      expect(formFields.length).toBe(1);
      expect(typeof firstFormField).toBe('object');
      expect(firstFormField).toMatchObject({ name: 'test' });
      expect(firstFormField).toMatchObject({ title: 'test title (optional)' });
      expect(firstFormField).toMatchObject({ value: 'test' });
      expect(firstFormField).toMatchObject({ type: 'string' });
      expect(firstFormField).toMatchObject({ validation: '' });
    });
  });
});
