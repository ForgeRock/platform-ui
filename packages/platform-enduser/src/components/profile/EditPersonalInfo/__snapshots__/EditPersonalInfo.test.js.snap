// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditPersonalInfo.vue EditPersonalInfo modal loaded 1`] = `
<bmodal-stub id="userDetailsModal" size="md" ignoreenforcefocusselector="" title="" titletag="h5" modalclass="fr-full-screen" headerclosecontent="&amp;times;" headercloselabel="Close" canceltitle="Cancel" oktitle="OK" cancelvariant="outline-secondary" okvariant="primary">
  <bcontainer-stub tag="div">
    <brow-stub tag="div">
      <bcol-stub sm="8" offsetsm="2" tag="div">
        <h3 class="text-center">
          No profile fields available
        </h3>
      </bcol-stub>
    </brow-stub>
  </bcontainer-stub>
</bmodal-stub>
`;
